crates/subspace-networking/src/behavior/custom_record_store.rs

```diff
@@ -242,6 +243,50 @@ impl<'a> RecordStorage<'a> for MemoryRecordStorage {
     }
 }
 
+// Workaround for Multihash::Sector until we fix https://github.com/libp2p/rust-libp2p/issues/3048
+// It returns `new_record` in case of other multihash or non-GSet values
+fn merge_records_in_case_of_sector_multihash(
+    new_record: Record,
+    old_record: Option<Record>,
+) -> Record {
+    let updated_rec = old_record.and_then(|old_record| {
+        let key_multihash = old_record.key.to_vec();
+
+        let multihash = Multihash::from_bytes(key_multihash.as_slice())
+            .expect("Key should represent a valid multihash");
+
+        if multihash.code() == u64::from(MultihashCode::Sector) {
+            let gset1 = if let Ok(gset) = GSet::<Vec<u8>>::decode(&mut old_record.value.as_slice())
+            {
+                gset
+            } else {
+                // Value is not a GSet
+                return Some(new_record.clone());
+            };
+
+            let gset2 = if let Ok(gset) =
+                GSet::<Vec<u8>>::decode(&mut new_record.value.clone().as_slice())
+            {
+                gset
+            } else {
+                // Value is not a GSet
+                return Some(new_record.clone());
+            };
+
+            let merged_gset = gset1.merge(gset2);
+
+            Some(Record {
+                value: merged_gset.encode(),
+                ..new_record.clone()
+            })
```

Comment:
Isn't this equivalent?:
```suggestion
            let mut value = HashSet::<Vec<u8>>::decode(&mut old_record.value.as_slice()).ok()?;
            value.extend(HashSet::<Vec<u8>>::decode(&mut new_record.value.as_slice()).ok()?;

            Some(Record {
                value: value.encode(),
                ..new_record.clone()
            })            
```

Reply:
It doesn't seem so. Suggested code could return None in different cases than the original code.

Reply:
It returns `None` in cases old code was returning `new_record.clone()` because in that case `new_record` will be used as a fallback (`updated_rec.unwrap_or(new_record)`) anyway, so no need to have extra code and unnecessary cloning.